drivers/input/input_gpio_qdec.c

```diff
@@ -0,0 +1,225 @@
+{
+	const struct gpio_qdec_config *cfg = dev->config;
+	unsigned int flags = enable ? GPIO_INT_EDGE_BOTH : GPIO_INT_DISABLE;
+	int ret;
+
+	for (int i = 0; i < 2; i++) {
+		const struct gpio_dt_spec *gpio = &cfg->gpio[i];
+
+		ret = gpio_pin_interrupt_configure_dt(gpio, flags);
+		if (ret != 0) {
+			LOG_ERR("Pin %d interrupt configuration failed: %d", i, ret);
+			return;
+		}
+	}
+}
+
+static void gpio_qdec_idle_worker(struct k_work *work)
+{
+	struct gpio_qdec_data *data = CONTAINER_OF(
+			work, struct gpio_qdec_data, idle_work);
+	const struct device *dev = data->dev;
+
+	k_timer_stop(&data->sample_timer);
+
+	gpio_qdec_irq_setup(dev, true);
+
+	LOG_DBG("polling stop");
+}
+
+static void gpio_qdec_cb(const struct device *gpio_dev, struct gpio_callback *cb,
+			 uint32_t pins)
+{
+	struct gpio_qdec_data *data = CONTAINER_OF(
+			cb, struct gpio_qdec_data, gpio_cb);
+	const struct device *dev = data->dev;
+	const struct gpio_qdec_config *cfg = dev->config;
+
+	gpio_qdec_irq_setup(dev, false);
+
+	k_timer_start(&data->sample_timer, K_NO_WAIT,
+		      K_USEC(cfg->sample_time_us));
+
+	LOG_DBG("polling start");
+}
+
+static int gpio_qdec_init(const struct device *dev)
+{
+	const struct gpio_qdec_config *cfg = dev->config;
+	struct gpio_qdec_data *data = dev->data;
+	int ret;
+
+	data->dev = dev;
+
+	gpio_pin_configure_dt(&cfg->gpio[0], GPIO_INPUT);
+	gpio_pin_configure_dt(&cfg->gpio[1], GPIO_INPUT);
+
+	k_work_init(&data->event_work, gpio_qdec_event_worker);
+	k_work_init_delayable(&data->idle_work, gpio_qdec_idle_worker);
+
+	k_timer_init(&data->sample_timer, gpio_qdec_sample_timer_timeout, NULL);
+	k_timer_user_data_set(&data->sample_timer, (void *)dev);
+
+	gpio_init_callback(&data->gpio_cb, gpio_qdec_cb,
+			   BIT(cfg->gpio[0].pin) | BIT(cfg->gpio[1].pin));
+	for (int i = 0; i < 2; i++) {
+		const struct gpio_dt_spec *gpio = &cfg->gpio[i];
+
+		if (!gpio_is_ready_dt(gpio)) {
+			LOG_ERR("%s is not ready", gpio->port->name);
+			return -ENODEV;
+		}
+
+		ret = gpio_pin_configure_dt(gpio, GPIO_INPUT);
+		if (ret != 0) {
+			LOG_ERR("Pin %d configuration failed: %d", i, ret);
+			return ret;
+		}
+
+		gpio_add_callback_dt(gpio, &data->gpio_cb);
+	}
+
+	gpio_qdec_irq_setup(dev, true);
+
+	LOG_DBG("Device %s initialized", dev->name);
+
+	return 0;
+}
+
+#define QDEC_GPIO_INIT(n)						\
+	static const struct gpio_qdec_config gpio_qdec_cfg_##n = {	\
+		.gpio = {GPIO_DT_SPEC_INST_GET_BY_IDX(n, gpios, 0),	\
+			 GPIO_DT_SPEC_INST_GET_BY_IDX(n, gpios, 1)},	\
+		.sample_time_us = DT_INST_PROP(n, sample_time_us),	\
+		.idle_timeout_ms = DT_INST_PROP(n, idle_timeout_ms),	\
+		.steps_per_period = DT_INST_PROP(n, steps_per_period),	\
+		.axis = DT_INST_PROP(n, zephyr_axis),			\
+	};								\
+									\
+	static struct gpio_qdec_data gpio_qdec_data_##n;		\
+									\
+	DEVICE_DT_INST_DEFINE(n, gpio_qdec_init, NULL,			\
+			      &gpio_qdec_data_##n,			\
+			      &gpio_qdec_cfg_##n,			\
+			      POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY,	\
```

Comment:
`SENSOR`?

Reply:
Leftover from this code previous life. :-) Fixed, thanks for spotting it.